### ISSUE TITLE
IfcOpenShell python viewer crashes on IfcRelNests

### ISSUE THREAD

**Issue by Moult** [IfcOpenShell/IfcOpenShell#681]
The python-occ geometry app started with:

```
from ifcopenshell.geom.app import application
application().start()
```

... crashes when attempting to open this file. The IfcRelNests line seems to trigger the crash:

```
#126=IFCRELNESTS('2icBcJgtj90RTpdmCB$2Ao',#5,$,$,#117,(#124));
```

If the line is removed, the file is loaded, and the nested element `#124` has geometry displayed (even though it probably shouldn't, as it is not part of the spatial tree).

Ideas? Or is there something wrong with the way I am doing the rel nests?

Sample file:

[output.ifc.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/3665247/output.ifc.txt)

**Comment by aothms** [comment]
It seems to work for me:

![afbeelding](https://user-images.githubusercontent.com/1096535/65814292-23bca400-e1e0-11e9-9572-731854310248.png)

What version of the app are you running?

**Comment by Moult** [comment]
I've compiled the latest v0.6.0 branch of ifcopenshell compiled against OCC 7.3.0. I'm running 03f31d5 which I think is 0.18.1 (https://github.com/tpaviot/pythonocc-core/commit/03f31d55dbcebb6c8aa23cffa978b22837b02467) of pythonocc-core. Maybe that's the issue? Should I upgrade?

**Comment by aothms** [comment]
> pythonocc-core. Maybe that's the issue? Should I upgrade?

Unlikely. PythonOCC is mainly for the visualization. Crashes are more likely during geometry processing.

> crash

Crash is an actual C++ segmentation fault or exception or just something in Python?

In case of the first. Would be great if you can run in gdb to get a stacktrace. Just `gdb python` and then the usual fluff should be enough to get a stacktrace.

**Comment by Moult** [comment]
Segmentation fault.

Here's my attempt. Apologies if I am doing this wrong:

```
$ gdb python
GNU gdb (Gentoo 8.3 vanilla) 8.3
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://bugs.gentoo.org/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from python...
(No debugging symbols found in python)
(gdb) run
Starting program: /usr/bin/python 
process 16432 is executing new program: /usr/bin/python3.7m
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Python 3.7.4 (default, Aug 29 2019, 21:08:40) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from ifcopenshell.geom.app import application
INFO:OCC.Display.backend:backend loaded: qt-pyqt5
>>> application().start()
[New Thread 0x7fffe1498700 (LWP 16480)]
[New Thread 0x7fffe0886700 (LWP 16481)]
[Detaching after fork from child process 16482]
QLayout: Attempting to add QLayout "" to code_edit "", which already has a layout
 ###### 3D rendering pipe initialisation #####
Display3d class initialization starting ...
Aspect_DisplayConnection created.
Graphic_Driver created.
V3d_Viewer created.
AIS_InteractiveContext created.
V3d_View created
Xw_Window created.
Display3d class successfully initialized.
 ########################################
Layer manager created
Layer dimensions: 477, 281
TKOpenGl | Type: Error | ID: 0 | Severity: High | Message:
  Failed to compile shader object. Compilation log:
0(265) : error C1503: undefined variable "occFragColor"

TKOpenGl | Type: Error | ID: 0 | Severity: High | Message:
  Error! FBO blitting has failed
[New Thread 0x7fffd3fff700 (LWP 16483)]
[Thread 0x7fffd3fff700 (LWP 16483) exited]
[New Thread 0x7fffd3fff700 (LWP 16485)]
property set dictionary has 0 entries
[New Thread 0x7fffd37fe700 (LWP 16486)]
[New Thread 0x7fffd2ffd700 (LWP 16487)]
[New Thread 0x7fffd27fc700 (LWP 16488)]
[New Thread 0x7fffd1ffb700 (LWP 16489)]
[Thread 0x7fffd27fc700 (LWP 16488) exited]
[Thread 0x7fffd1ffb700 (LWP 16489) exited]
[Thread 0x7fffd37fe700 (LWP 16486) exited]
Thread 7 "geometry_creati" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffd2ffd700 (LWP 16487)]
0x00007ffff62730ac in IfcGeom::Kernel::get_decomposing_entity(IfcUtil::IfcBaseEntity*, bool) () from /usr/lib/python3.7/site-packages/ifcopenshell/_ifcopenshell_wrapper.so
(gdb) bt
#0  0x00007ffff62730ac in IfcGeom::Kernel::get_decomposing_entity(IfcUtil::IfcBaseEntity*, bool) () from /usr/lib/python3.7/site-packages/ifcopenshell/_ifcopenshell_wrapper.so
#1  0x00007ffff631c898 in IfcGeom::BRepElement<double, double>* IfcGeom::KernelIfc4::create_brep_for_representation_and_product<double, double>(IfcGeom::IteratorSettings const&, Ifc4::IfcRepresentation*, Ifc4::IfcProduct*) ()
   from /usr/lib/python3.7/site-packages/ifcopenshell/_ifcopenshell_wrapper.so
#2  0x00007ffff633692d in void (anonymous namespace)::create_element<double, double>(IfcGeom::KernelIfc4*, IfcGeom::IteratorSettings const&, (anonymous namespace)::geometry_conversion_task<double, double>*) ()
   from /usr/lib/python3.7/site-packages/ifcopenshell/_ifcopenshell_wrapper.so
#3  0x00007ffff63234c2 in std::_Function_handler<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> (), std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, std::thread::_Invoker<std::tuple<void (*)(IfcGeom::KernelIfc4*, IfcGeom::IteratorSettings const&, (anonymous namespace)::geometry_conversion_task<double, double>*), IfcGeom::KernelIfc4*, std::reference_wrapper<IfcGeom::IteratorSettings>, (anonymous namespace)::geometry_conversion_task<double, double>*> >, void> >::_M_invoke(std::_Any_data const&) () from /usr/lib/python3.7/site-packages/ifcopenshell/_ifcopenshell_wrapper.so
#4  0x00007ffff62ba089 in std::__future_base::_State_baseV2::_M_do_set(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>*, bool*) ()
   from /usr/lib/python3.7/site-packages/ifcopenshell/_ifcopenshell_wrapper.so
#5  0x00007ffff7c6c357 in __pthread_once_slow () from /lib64/libpthread.so.0
#6  0x00007ffff632493b in std::thread::_State_impl<std::thread::_Invoker<std::tuple<std::__future_base::_Async_state_impl<std::thread::_Invoker<std::tuple<void (*)(IfcGeom::KernelIfc4*, IfcGeom::IteratorSettings const&, (anonymous namespace)::geometry_conversion_task<double, double>*), IfcGeom::KernelIfc4*, std::reference_wrapper<IfcGeom::IteratorSettings>, (anonymous namespace)::geometry_conversion_task<double, double>*> >, void>::_Async_state_impl(std::thread::_Invoker<std::tuple<void (*)(IfcGeom::KernelIfc4*, IfcGeom::IteratorSettings const&, (anonymous namespace)::geometry_conversion_task<double, double>*), IfcGeom::KernelIfc4*, std::reference_wrapper<IfcGeom::IteratorSettings>, (anonymous namespace)::geometry_conversion_task<double, double>*> >&&)::{lambda()#1}> > >::_M_run() () from /usr/lib/python3.7/site-packages/ifcopenshell/_ifcopenshell_wrapper.so
#7  0x00007ffff1aff60e in std::execute_native_thread_routine (__p=0x7fffcc06a380) at /var/tmp/portage/sys-devel/gcc-8.3.0-r1/work/gcc-8.3.0/libstdc++-v3/src/c++11/thread.cc:80
#8  0x00007ffff7c63508 in start_thread () from /lib64/libpthread.so.0
#9  0x00007ffff7b90bff in clone () from /lib64/libc.so.6
(gdb)
```

**Comment by aothms** [comment]
No this is perfect.

`IfcGeom::Kernel::get_decomposing_entity()` so it really has to do with the element hierarchy. Nothing version or os specific it seems. Strange that I wasnt able to reproduce. I'll try harder using the exact same commit sha you mentioned.

**Comment by Moult** [comment]
A small note that theoretically this line:

```
#122=IFCLOCALPLACEMENT($,#121);
```

... should be ...

```
#122=IFCLOCALPLACEMENT(#115,#121);
``` 

... due to the rules described on [this page](https://standards.buildingsmart.org/IFC/RELEASE/IFC4/ADD2_TC1/HTML/link/ifclocalplacement.htm), but it seems unrelated to the crash.

**Comment by Moult** [comment]
Interestingly, the IfcOpenShell compiled by the build bot does not crash, but the IfcOpenShell compiled by myself does crash. I found that if I copied the ifcopenshell directory from the build bot package to my site-packages, it can open the file successfully.

Perhaps a related issue: this test.ifc file opens in the ifcopenshell compiled by myself (latest v0.6.0 branch) but it shows one cube, whereas it should show two cubes. Perhaps it is a regression since the build bot version, or perhaps something funny with my build environment.

[test.ifc.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/3670918/test.ifc.txt)

My compilation steps on Linux are as follows:

```
$ cd IfcOpenShell && mkdir build && cd build
$ cmake ../cmake -DCOLLADA_SUPPORT=0 -DIFCXML_SUPPORT=off -DOCC_INCLUDE_DIR="/usr/lib64/opencascade-7.3.0/ros/include/opencascade" -DOCC_LIBRARY_DIR="/usr/lib64/opencascade-7.3.0/ros/lib64"
$ make -j12
$ sudo make install
```
